js/src/common/components/LoadingIndicator.tsx

```diff
@@ -40,21 +42,30 @@ export interface LoadingIndicatorAttrs extends ComponentAttrs {
  *
  * - `containerClassName` Class name(s) to apply to the indicator's parent
  * - `className` Class name(s) to apply to the indicator itself
- * - `size` Size of the loading indicator
+ * - `size` Size of the loading indicator (small, medium or large)
  * - `containerAttrs` Optional attrs to be applied to the container DOM element
+ * - `block` Displays the spinner as a block element with a height of 100px
+ * - `inline` Displays the spinner as an inline element, centrally aligned vertically.
  *
  * All other attrs will be assigned as attributes on the DOM element.
  */
 export default class LoadingIndicator extends Component<LoadingIndicatorAttrs> {
   view() {
-    const { size, ...attrs } = this.attrs;
+    const { inline = false, block = false, size = 'medium', containerClassName, className, ...attrs } = this.attrs;
 
-    attrs.className = classList({ LoadingIndicator: true, [attrs.className || '']: true });
-    attrs.containerClassName = classList({ 'LoadingIndicator-container': true, [attrs.containerClassName || '']: true });
+    if (inline && block) throw new Error('LoadingIndicator cannot have both `inline` and `block` attributes.');
+
+    const completeClassName = classList('LoadingIndicator', className);
+    const completeContainerClassName = classList(
+      'LoadingIndicator-container',
+      block && 'LoadingIndicator--block',
+      inline && 'LoadingIndicator--inline',
+      containerClassName
+    );
 
     return (
-      <div {...attrs.containerAttrs} data-size={size} className={attrs.containerClassName}>
-        <div {...attrs}></div>
+      <div {...attrs.containerAttrs} data-size={size} className={completeContainerClassName}>
```

Comment:
a bit out of scope of this PR, but any reason for using `data-*` instead of a modifier class which is more in line with convention ?

Reply:
Data attributes are intended as a private storage area for DOM elements, and I prefer using them wherever possible for things like this.

They're slightly less performant than classes (9000 operations/sec vs 13000 ops/s), but I find `data-size="large"` more verbose and clear than `class="LoadingIndicator LoadingIndicator--large"`. The change in performance is basically negligible -- it's still 9000 operations per second!

Reply:
> Data attributes are intended as a private storage area for DOM elements, and I prefer using them wherever possible for things like this.

Hmm, I still think a modifier class is more appropriate and correct, as per https://johnresig.com/blog/html-5-data-attributes/ data attributes aren't meant to affect layout or presentation, we're only using `size` here to change the way the loading indicator is presented, so I think we're misusing it.

Reply:
That blog post talks about the fact that data attributes aren't used by the browser to adjust layout and presentation, like other attributes can be (e.g. deprecated `align`, `hidden`, etc). Maybe we should get some more opinions on this?

Reply:
> Places such as CSS Tricks seem to show them being used for things like this

I'm not completely opposed to using it, there are valid use cases, I just don't see the benefit of using it over a class when the sole purpose is styling, especially as the rest of the code base already uses a convention in which modifier classes are used.

Additionally there is a reason for the current convention we follow, one that I find very valuable with a customizable software. Specificity, a `LoadingIndicator--small` is less specific than `LoadingIndicator[data-size="small"]`, and having tried multiple times customizing the software, specificity was always the most tedious thing to deal with, as you have to overspecify even more or even use `!important`.

It's best if we figure this out now 🤔

Reply:
You're right -- it has higher specificity, but that's not inherently wrong.

The main reason I like data attributes is because it makes it clear what data can be used to modify the element and what that actually means. Most of the time, we won't need to customise the size, so developers may only see the `LoadingIndicator` class and never the size classes, meaning that they may not be aware that they exist.

If we have `LoadingIndicator--small` (and other sizes), then `LoadingIndicator--inline` (and block), it starts becoming unclear what the modifiers are actually representing. It's not a massive issue in this component, but could be in others.

Reply:
I think I prefer using classes. I strongly doubt that this is the bottleneck for performance, but better customizability is very important. Also, I don't see why block/inline should be classes, but not small/medium/large.

Alternatively, maybe we could set it as a css variable?

Reply:
In this particular use case I believe that using `LoadingIndicator--{size}` would be correct to meet our existing standards and uses.

Reply:
We can't really set it as a CSS variable, but we do have support for manually specifying size as variables using `--size: 36px` (for example).

I would have personally chosen to use `data-display` with `block` and `inline`, but I don't see any other display options that we might need.

I see the size as more of a variable than anything, which splitting into an attribute makes it more similar to normal HTML elements, such as `<input type="number">`, for example.

Reply:
Matt's comment disappeared into thin air when I rebased onto master, so I copy-pasted it from the email notification:

> In this particular use case I believe that using `LoadingIndicator--{size}` would be correct to meet our existing standards and uses.